src/libraries/Common/src/SourceGenerators/TypeModelHelper.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text;
+using Microsoft.CodeAnalysis;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace SourceGenerators
+{
+    internal static class TypeModelHelper
+    {
+        private static readonly SymbolDisplayFormat s_minimalDisplayFormat = new SymbolDisplayFormat(
+            globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+            typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypes,
+            genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+            miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+
+        public static string ToIdentifierCompatibleSubstring(this ITypeSymbol type, bool useUniqueName)
```

Comment:
I don't think this algorithm should live in shared code. It should be specific to each application since changing the algorithm in the future is a potential breaking change, and what we have here isn't perfect.

Reply:
Okay I'll revert in this PR. 

---

For 9.0, I still think we should at least have the separate implementations live next to each other so they can be kept in sync. If one generator makes an important improvement, then we would address the other one if needed.

IMO this would also increase discoverability of common concerns when handling type graphs in source-gen (formatting identifiers, visibility checks, formatting names with special syntax e.g. `@int` etc.) Authors would know to check the shared code dir for helper APIs that already exist and they might not need to rewrite.